app/controllers/bulk_downloads_controller.rb

```diff
@@ -75,6 +93,69 @@ def show
     render json: { error: BulkDownloadsHelper::BULK_DOWNLOAD_NOT_FOUND }, status: :not_found
   end
 
+  # GET /bulk_downloads/:id/presigned_output_url
+  def presigned_output_url
+    bulk_download_id = params[:id]
+
+    bulk_download = current_power.viewable_bulk_downloads.find(bulk_download_id)
+
+    if bulk_download.status != BulkDownload::STATUS_SUCCESS
+      render json: { error: BulkDownloadsHelper::OUTPUT_FILE_NOT_SUCCESSFUL }, status: :not_found
+      return
+    end
+
+    presigned_url = bulk_download.output_file_presigned_url
+
+    if presigned_url.nil?
+      render json: { error: BulkDownloadsHelper::PRESIGNED_URL_GENERATION_ERROR }, status: :internal_server_error
+      return
+    end
+    render json: presigned_url
+  rescue ActiveRecord::RecordNotFound
+    render json: { error: BulkDownloadsHelper::BULK_DOWNLOAD_NOT_FOUND }, status: :not_found
+  end
+
+  # POST /bulk_downloads/:id/success/:access_token
+  def success_with_token
+    # set bulk download and validate access token in before_action
+    @bulk_download.update(status: BulkDownload::STATUS_SUCCESS)
+
+    if params[:error_type] == "FailedSrcUrlError"
+      LogUtil.log_err_and_airbrake("BulkDownloadFailedSrcUrlError: The following paths failed to process: #{params[:error_data]}")
+      @bulk_download.update(error_message: FAILED_SRC_URL_ERROR_TEMPLATE % params[:error_data].length)
+    end
+
+    render json: { status: "success" }
+  end
+
+  # POST /bulk_downloads/:id/error/:access_token
+  def error_with_token
+    # set bulk download and validate access token in before_action
+    @bulk_download.update(status: BulkDownload::STATUS_ERROR, error_message: params[:error_message])
+    LogUtil.log_err_and_airbrake("BulkDownloadFailedError: #{params[:error_message]}")
+    render json: { status: "success" }
+  end
+
+  # POST /bulk_downloads/:id/progress/:access_token
+  def progress_with_token
+    # set bulk download and validate access token in before_action
+    @bulk_download.update(progress: params[:progress].to_f)
+    render json: { status: "success" }
+  end
+
+  def set_bulk_download_from_id
+    @bulk_download = BulkDownload.find(params[:id])
```

Comment:
this should really use `current_power.viewable_bulk_downloads.find`, right? because some other methods assume that @bulk_download is something you are authorized to use

Reply:
I combined this method with the one below to make it less ambiguous. This one is for endpoints using the access token authentication, so there isn't a current user in this case.